composer/algorithms/factorize/factorize_core.py

```diff
@@ -0,0 +1,353 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+import dataclasses
+from typing import Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+
+@dataclasses.dataclass
+class LowRankSolution:
+    """Bundles tensors used by a factorized linear operator.
+
+    The factorization always splits the operator into two smaller linear
+    operators. The first takes in input of the original shape and embeds it
+    in a lower-dimensional space. The second maps this lower-dimensional space
+    to the original output space.
+
+    Args:
+        Wa: First linear operation in the factorized approximation. For a
+            factorized linear operation, ``Wa`` is a matrix. For a factorized
+            convolution, ``Wa`` matches the shape of the convolution's
+            original weight parameter, except along the channel axis.
+        Wb: Second linear operation in the factorized approximation. Shape
+            is such that composing ``Wb`` with ``Wb`` yields an output of
+            the same size as the original operation.
+        bias: vector added to the output of the second linear operation
+        rank: output dimensionality (channels or features) of the first linear
+            operation, and input dimensionality of the second input operation.
+        nmse: normalized mean squared error obtained during the optimization
+            procedure used to derive ``Wa``, ``Wb``, and ``bias``. This is
+            equal to the raw mean squared error between the factorized
+            approximation's output and the original output, divided by the
+            variance of the original output. A value of 0 means no error
+            was introduced, and a value of 1 corresponds to capturing the
+            output no better than chance.
+    """
+    Wa: Optional[torch.Tensor] = None
+    Wb: Optional[torch.Tensor] = None
+    bias: Optional[torch.Tensor] = None
+    rank: int = -1
+    nmse: float = 0
+
+
+def _lstsq(A: torch.Tensor, B: torch.Tensor) -> torch.Tensor:
+    if A.shape[0] != B.shape[0]:
+        raise IndexError("A has different number of rows than B! " f"A.shape = {A.shape}, B.shape = {B.shape}")
```

Comment:
```suggestion
        raise ValueError("A has different number of rows than B! " f"A.shape = {A.shape}, B.shape = {B.shape}")
```
I think `IndexError` should only be raised when slicing. See https://docs.python.org/3/library/exceptions.html#IndexError